### DISCUSSION TITLE
Network adapter order

### DISCUSSION THREAD

**Discussion by petrasl1976** [ipxe/ipxe#333] [Q&A]
Hi,

i found the post https://forum.ipxe.org/showthread.php?tid=7337&pid=10525#pid10525 where `robinsmidsrod` says that linux and iPXE uses different method to enumerate PCI buss, devices. While i am in linux (lets say debian) how can i predict which mac will be for net0, net1, ... looking from iPXE perspective?

I install iPXE to disk and reboot server from it, so iPXE starts and downloads kernel and initrd from the network. I compile ip, gw and mask statically, so i also would like to know which device is wired. If i miss and this is not net0 - i need to boot server from media and reinstall iPXE by providing other device (net1). Sometimes i have servers with 2x4 adapters and cable connected not necessary to the firs port.

There is a way to go through all adapters, by setting them and try to ping GW, but it takes extra time and i do not want to delay server boot at this point.

I would appreciate if someone would describe or show how to list network adapters the way iPXE would list them.

Thanks

**Comment by mcb30** [comment]
For BIOS versions of iPXE, devices will always be enumerated in increasing order of PCI bus:dev.fn address.

For UEFI versions of iPXE, the enumeration order is entirely up to the platform firmware and iPXE has no control over it.

You can use the MAC address as a persistent identifier: e.g. iterate over all network devices in an iPXE script and check `${net${idx}/mac}` to find your wired NIC, then use `set net${idx}/ip .....` etc to set the static IP address.

You may also want to consider using an `autoexec.ipxe` script loaded from the disk, so that you don't have to recompile iPXE each time.

**Reply by petrasl1976** [reply]
So you think something like that should work (full script below)? But it doesn't :( I have tried it already and i think because of `iseq ...` command. ... or i do not know how to compare mac and my preferable string?
```
#!ipxe
imgfree
set idx:int32 0
:loop isset ${net${idx}/mac} || goto reboot
  echo net${idx}/mac = ${net${idx}/mac}
  iseq ${net${idx}/mac} 5E:8A:AC:31:02:4C && goto setnet ||
  inc idx && goto loop

:setnet
set net${idx}/ip 192.168.1.2
set net${idx}/netmask 255.255.255.255
set net${idx}/gateway 192.168.1.1
ifopen net${idx} && goto boot || goto reboot

:boot
initrd https://mytftp.company.org/initrd || goto reboot
kernel https://mytftp.company.org/vmlinuz ip=192.168.1.2::192.168.1.1:255.255.255.0::enp2s0:none:1.1.1.1 boot=live components cgroup_enable=memory swapaccount=1 ipv6.disable=1 fetch=https://mytftp.company.org/squashfs --- || goto reboot
module https://mytftp.company.org/v1/boot/?state=ipxe_ok || echo Cant report boot state. Continue...
boot vmlinuz || goto reboot

:reboot
echo Unable to get or boot vmlinuz.
prompt --key 0x02 --timeout 10000 Press Ctrl-B for the iPXE command line... && shell || reboot
reboot
```

Its not a problem to recompile and rewrite to disk. My cattle use disks just for iPXE, everything runs in memory. My issue: to know exact name of net adapter in iPXE while i am running Linux :)

Thanks!

**Reply by mcb30** [reply]
`iseq` is case-sensitive, and MAC addresses in iPXE will be lower-case.  You should find that
```
iseq ${net${idx}/mac} 5e:8a:ac:31:02:4c && goto setnet ||
```
will work.

**Reply by petrasl1976** [reply]
Oh man... "And twelve points go to theee... " :D
Thanks! That was my issue
